File: LU_factorization_pivot.py
import numpy as np

class LUSolver:
    def __init__(self, size):
        self.n = size
        self.A = np.zeros((size, size))
        self.b = np.zeros(size)
        self.L = np.zeros((size, size))
        self.U = np.zeros((size, size))
        self.pivot = np.arange(size)

    def set_matrix(self, matrix):
        self.A = matrix

    def set_vector(self, vector):
        self.b = vector

    def print_matrix(self, mat):
        for row in mat:
            print(" ".join(f"{val:10.4f}" for val in row))

    def print_vector(self, vec):
        for i, val in enumerate(vec):
            print(f"b[{i}] = {val:10.4f}")

    def decompose(self):
        n = self.n
        for i in range(n):
            maxRow = np.argmax(np.abs(self.A[i:, i])) + i
            if maxRow != i:
                self.A[[i, maxRow]] = self.A[[maxRow, i]]
                self.pivot[[i, maxRow]] = self.pivot[[maxRow, i]]

            for j in range(i, n):
                self.U[i, j] = self.A[i, j] - np.dot(self.L[i, :i], self.U[:i, j])

            for j in range(i, n):
                if i == j:
                    self.L[i, i] = 1
                else:
                    self.L[j, i] = (self.A[j, i] - np.dot(self.L[j, :i], self.U[:i, i])) / self.U[i, i]

    def forward_substitution(self):
        n = self.n
        y = np.zeros(n)
        for i in range(n):
            y[i] = self.b[self.pivot[i]] - np.dot(self.L[i, :i], y[:i])
        return y

    def back_substitution(self, y):
        n = self.n
        x = np.zeros(n)
        for i in range(n - 1, -1, -1):
            x[i] = (y[i] - np.dot(self.U[i, i + 1:], x[i + 1:])) / self.U[i, i]
        return x

    def solve(self):
        self.decompose()

        print("L matrix:")
        self.print_matrix(self.L)

        print("U matrix:")
        self.print_matrix(self.U)

        y = self.forward_substitution()

        print("Vector y after forward substitution:")
        self.print_vector(y)

        x = self.back_substitution(y)

        print("The solution to the system of equations is:")
        for i in range(self.n):
            print(f"x[{i}] = {x[i]:.4f}")
        return x

def main():
    n = 3
    solver = LUSolver(n)

    a = np.array([[1, 2, -1],
                  [-2, 3, 1],
                  [4, -1, -3]])
    b_values = np.array([-1, 0, -2])

    solver.set_matrix(a)
    solver.set_vector(b_values)

    print("Initial matrix:")
    solver.print_matrix(solver.A)

    print("Initial vector b:")
    solver.print_vector(solver.b)

    solver.solve()

if __name__ == "__main__":
    main()


"""
Explanation:
LUSolver Class:

Manages the matrix 
𝐴
A, vector 
𝑏
b, L and U matrices, and the pivot array.
Methods to set the matrix and vector (set_matrix and set_vector).
Methods to print matrices and vectors (print_matrix and print_vector).
The decompose method performs LU decomposition with pivoting.
The forward_substitution and back_substitution methods solve the system of equations.
The solve method orchestrates the solving process, including decomposition and printing intermediate steps.
Main Function:

Initializes the LUSolver with matrix 
𝐴
A and vector 
𝑏
b.
Prints the initial matrix and vector.
Calls the solve method to perform LU decomposition, solve the system, and print the solution.
This implementation consolidates all functionality within the LUSolver class, making the code more organized and easier to manage.
"""

File: LU_factorization_pivot.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
// Author: Vasile-Daniel DAN 
// Start
// Project: LU Factorization - Doolittle (WITH pivoting)
//////////////////////////////////////////////////////////////////////////////////  
// Functions for printing matrices and arrays
void printMatrix(int n, double matrix[n][n]) {
    int i, j;
    for (i = 0; i < n; i++) {
        for (j = 0; j < n; j++) {
            printf("%10.4lf ", matrix[i][j]);
        }
        printf("\n");
    }
}

void printArray(int n, double arr[n]){
    int i;
    for (i = 0; i < n; i++) {
        printf("b[%d] = %10.4lf \n", i, arr[i]);
    }
}

// Function to swap two rows in a matrix
void swapRows(int n, double matrix[n][n], int row1, int row2) {
    double temp;
    for (int i = 0; i < n; i++) {
        temp = matrix[row1][i];
        matrix[row1][i] = matrix[row2][i];
        matrix[row2][i] = temp;
    }
}

// Function to swap two elements in an array
void swapElements(double arr[], int i, int j) {
    double temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

// LU decomposition function (Doolittle variant) with pivoting
void luDecomposition(int n, double a[n][n], double l[n][n], double u[n][n], int pivot[n]) {
    int i, j, k;
    for (i = 0; i < n; i++) {
        pivot[i] = i;
    }

    for (i = 0; i < n; i++) {
        // Partial pivoting
        double max = fabs(a[i][i]);
        int maxRow = i;
        for (k = i + 1; k < n; k++) {
            if (fabs(a[k][i]) > max) {
                max = fabs(a[k][i]);
                maxRow = k;
            }
        }

        if (maxRow != i) {
            swapRows(n, a, i, maxRow);
            swapElements(pivot, i, maxRow);
        }

        // Initialize U
        for (j = i; j < n; j++) {
            u[i][j] = a[i][j];
            for (k = 0; k < i; k++) {
                u[i][j] -= l[i][k] * u[k][j];
            }
        }

        // Initialize L
        for (j = i; j < n; j++) {
            if (i == j) {
                l[i][i] = 1;
            } else {
                l[j][i] = a[j][i];
                for (k = 0; k < i; k++) {
                    l[j][i] -= l[j][k] * u[k][i];
                }
                l[j][i] /= u[i][i];
            }
        }
    }
}

// Forward substitution to solve Ly = Pb
void forwardSubstitution(int n, double l[n][n], double b[n], double y[n], int pivot[n]) {
    for (int i = 0; i < n; i++) {
        y[i] = b[pivot[i]];
        for (int j = 0; j < i; j++) {
            y[i] -= l[i][j] * y[j];
        }
    }
}

// Backward substitution to solve Ux = y
void backSubstitution(int n, double u[n][n], double y[n], double x[n]) {
    for (int i = n - 1; i >= 0; i--) {
        x[i] = y[i];
        for (int j = i + 1; j < n; j++) {
            x[i] -= u[i][j] * x[j];
        }
        x[i] /= u[i][i];
    }
}

int main() {
    int n = 3;
    double x[3], y[3];
    int pivot[3];

    double a[3][3] = {
        { 1, 2, -1},
        {-2, 3, 1},
        { 4, -1, -3}
    };

    printf("Initial matrix:\n");
    printMatrix(n, a);

    double b[3] = {-1, 0, -2};

    printf("Initial vector b:\n");
    printArray(n, b);

    double l[3][3] = {0}, u[3][3] = {0};

    luDecomposition(n, a, l, u, pivot);

    printf("L matrix:\n");
    printMatrix(n, l);
    printf("U matrix:\n");
    printMatrix(n, u);

    forwardSubstitution(n, l, b, y, pivot);

    printf("Vector y after forward substitution:\n");
    printArray(n, y);

    backSubstitution(n, u, y, x);

    printf("The solution to the system of equations is:\n");
    for(int i = 0; i < n; i++){
        printf("x[%d] = %lf \n", i, x[i]);
    }

    return 0;
}
